# origin: mlafon (BitKeeper)
# cset: 1.1982.161.36 (2.6) key=4244bfc9vHVlT4nv2o4ys4_sf6vzKA
# URL: http://linux.bkbits.net:8080/linux-2.6/cset@4244bfc9vHVlT4nv2o4ys4_sf6vzKA
# inclusion: upstream
# descrition: [PATCH] Suspected information leak (mem pages) in ext2
# revision date: Mon, 28 Mar 2005 18:40:00 +0900
#
# S rset: ChangeSet|1.1982.161.35..1.1982.161.36
# I rset: fs/ext2/dir.c|1.27..1.28
#
# Key:
# S: Skipped  ChangeSet file only
# O: Original Followed by Updated
# U: Updated  Included with updated range of versions
# I: Included Included verbatim
# E: Excluded Excluded on request from user
# D: Deleted  Manually deleted by subsequent user edit
# R: Revised  Manually revised by subsequent user edit
#
#
# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2005/03/25 17:50:01-08:00 mlafon@arkoon.net 
#   [PATCH] Suspected information leak (mem pages) in ext2
#   
#   From: "Mathieu Lafon" <mlafon@arkoon.net>
#   
#   I think I have discovered a potential security problem in ext2: when a
#   new directory is created, the ext2 block written to disk is not
#   initialized.
#   
#   Included is a proposed patch for Linux 2.6 (ext2_make_empty() function):
#   
#   CAN-2005-0400 is assigned to this issue.
#   
#   Signed-off-by: Chris Wright <chrisw@osdl.org>
# 
# fs/ext2/dir.c
#   2005/03/16 10:17:01-08:00 mlafon@arkoon.net +1 -0
#   Suspected information leak (mem pages) in ext2
# 
#
===== fs/ext2/dir.c 1.27 vs 1.28 =====
--- 1.27/fs/ext2/dir.c	2005-01-05 11:48:14 +09:00
+++ 1.28/fs/ext2/dir.c	2005-03-17 03:17:01 +09:00
@@ -592,6 +592,7 @@ int ext2_make_empty(struct inode *inode,
 		goto fail;
 	}
 	kaddr = kmap_atomic(page, KM_USER0);
+       memset(kaddr, 0, chunk_size);
 	de = (struct ext2_dir_entry_2 *)kaddr;
 	de->name_len = 1;
 	de->rec_len = cpu_to_le16(EXT2_DIR_REC_LEN(1));
